diesel/src/sqlite/connection/raw.rs

```diff
@@ -292,99 +343,155 @@ extern "C" fn run_aggregator_step_function<ArgsSqlType, RetSqlType, Args, Ret, A
         // allocations occur, a null pointer is returned in this case
         // See: https://www.sqlite.org/c3ref/aggregate_context.html
         //
         // For the reasoning about the safety of the OptionalAggregator handling
         // see the comment in run_aggregator_step_function.
-        let aggregate_context = ffi::sqlite3_aggregate_context(ctx, 0);
-        let mut aggregate_context = NonNull::new(aggregate_context as *mut OptionalAggregator<A>);
-        let aggregator = match aggregate_context {
-            Some(ref mut a) => match std::mem::replace(a.as_mut(), OptionalAggregator::None) {
-                OptionalAggregator::Some(agg) => Some(agg),
-                OptionalAggregator::None => unreachable!("We've written to the aggregator in the xStep callback. If xStep was never called, then ffi::sqlite_aggregate_context() would have returned a NULL pointer")
-            },
-            None => None,
-        };
-
-        let result = A::finalize(aggregator);
+        ffi::sqlite3_aggregate_context(ctx, 0)
+    };
 
-        match process_sql_function_result::<RetSqlType, Ret>(result) {
-            Ok(value) => value.result_of(ctx),
-            Err(e) => {
-                let msg = e.to_string();
-                ffi::sqlite3_result_error(ctx, msg.as_ptr() as *const _, msg.len() as _);
+    let mut aggregate_context = NonNull::new(aggregate_context as *mut OptionalAggregator<A>);
+    let aggregator = aggregate_context.as_mut().map(|a| {
+        let a = unsafe { a.as_mut() };
+        match std::mem::replace(a, OptionalAggregator::None) {
+            OptionalAggregator::Some(agg) => agg,
+            OptionalAggregator::None => {
+                assert_fail!("We've written to the aggregator in the xStep callback. If xStep was never called, then ffi::sqlite_aggregate_context() would have returned a NULL pointer.");
             }
         }
+    });
+
+    let result = std::panic::catch_unwind(|| A::finalize(aggregator))
+        .map(process_sql_function_result::<RetSqlType, Ret>);
+
+    match result {
+        Ok(Ok(value)) => value.result_of(ctx),
+        Ok(Err(e)) => {
+            let msg = e.to_string();
+            unsafe { context_error_str(ctx, &msg) };
+        }
+        Err(_) => {
+            let msg = format!("{}::finalize() panicked", std::any::type_name::<A>());
+            unsafe { context_error_str(ctx, &msg) };
+        }
     }
 }
 
 unsafe fn null_aggregate_context_error(ctx: *mut ffi::sqlite3_context) {
     static NULL_AG_CTX_ERR: &str = "An unknown error occurred. sqlite3_aggregate_context returned a null pointer. This should never happen.";
 
-    ffi::sqlite3_result_error(
-        ctx,
-        NULL_AG_CTX_ERR.as_ptr() as *const _ as *const _,
-        NULL_AG_CTX_ERR.len() as _,
-    );
+    context_error_str(ctx, NULL_AG_CTX_ERR)
+}
+
+unsafe fn context_error_str(ctx: *mut ffi::sqlite3_context, error: &str) {
```

Comment:
We should cite https://sqlite.org/c3ref/result_blob.html here, stating that sqlite will copy our error message, so it's fine to free it afterwards. (Just to make that clear for any future person looking at this code.) 
Also we maybe can take a `NonNull<ffi::sqlite3_context>` here to make this function not `unsafe` to use anymore?